### PR TITLE
Check for self assign from numpy and copy

### LINKED ISSUE #1267: Assignment from numpy array fails to check for self-assignment
```python
import numpy as np
import scipp as sc
var = sc.Variable(dims=['x'], values=np.arange(100))
a = np.flip(var.values)
var.values = a
```

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #1267.

This is actually quite tricky with negative strides (in multiple dims), so I am not 100% certain my tests cover all cases. Please review carefully.

### CODE REVIEW

File: python/numpy.h [hunk 1]
```diff
@@ -61,11 +63,38 @@ void copy_flattened_4d(const py::array_t<T> &data, View &&view) {
...
+  const auto view_end = &*view.end();
+  // Note the use of std::less, pointer comparison with operator< may be
+  // undefined behavior with pointers from different arrays.
+  return std::less<const T *>()(data_begin, view_end) &&
+         std::greater_equal<const T *>()(data_end, view_begin);
+}
+
 template <class T, class View>
-void copy_flattened(const py::array_t<T> &data, View &&view) {
+void copy_flattened(py::array_t<T> data, View &&view) {
```
Comment (OwenArnold) [inline python/numpy.h:92]: Why not continue to take `data` as const ref?

Comment (SimonHeybrock) [reply python/numpy.h:92]: Because I may want to assign to it (a couple lines lower down).

Comment (OwenArnold) [reply python/numpy.h:92]: Discussed. `py::array<T>` is handle. This does not need to be changed.

File: python/numpy.h [hunk 2]
```diff
@@ -61,11 +63,38 @@ void copy_flattened_4d(const py::array_t<T> &data, View &&view) {
...
+bool memory_overlaps(const py::array_t<T> &data, const View &view) {
+  const auto &buffer_info = data.request();
+  const auto [data_begin, data_end] = memory_begin_end<T>(buffer_info);
+  const auto view_begin = &*view.begin();
+  const auto view_end = &*view.end();
+  // Note the use of std::less, pointer comparison with operator< may be
+  // undefined behavior with pointers from different arrays.
+  return std::less<const T *>()(data_begin, view_end) &&
+         std::greater_equal<const T *>()(data_end, view_begin);
+}
```
Comment (OwenArnold) [inline python/numpy.h:89]: Logic here looks good. Is this worth implementing and testing on the c++ side too? Or too hard to separate from `py::buffer_info` etc?

Comment (SimonHeybrock) [reply python/numpy.h:89]: Depends, what would we gain from testing such an implementation detail? Is there anything we cannot reasonably cover with the tests on the Python side?

**Pull request merged**